### ISSUE TITLE
zasm optimization: automatic CALL -> RST

### ISSUE THREAD

**Issue by hsoft** [hsoft/collapseos#72]
Efficiently use 0x08, 0x10, .. 0x38 has been on the back of my mind for a while and I've always pushed back to later. Reading Tiny Basic code and see mentions of RST brought back the issue, but the problem is always the same: how to decide what to put into those RST slots? Once apps begin targeting a particular slot with RST, it becomes harder to change for a user who would like to put another more important routine there.

So I had this idea: what if zasm had an optimisation in unconditional CALLs and that whenever NN was 0x08, 0x10 .. 0x38, it would spit the corresponding RST bytecode instead of the CALL?

This would allow a user to transparently put whatever she wishes into that zone and carefully craft a jump table in her `user.h` file that have the proper labels pointing to hotspots. All userspace apps would then benefit from free optimisations depending on the jump table crafted by the user.

It seems like a really nice idea to me but I wonder if someone can think of a drawback or if it could possibly introduce bugs.

**Comment by Allisontheolder** [comment]
most z80 uses shun that.  it also gets in the way of cheap and dirty 
interrupt or RST7 (used often for debuggers due to its code FFH which 
is typical of what many systems get off the bus if there is no ram.

Reason is interrupts and short calls while seemingly efficient really saves little 
other than a few bytes.  But they force code in in low space where rom may be 
and reset (RST-0) is kind of reserved.

Allison

**Comment by hsoft** [comment]
Insignificant? I don't know. It's not only two bytes saved, but also 6 T-states. Of course, the RST 0x38 slot is usually reserved, but there still 6 other slots to use.

**Comment by Allisontheolder** [comment]
Wow a few bytes.  It would be better to refactor the code and its utilization as then you can save a lot more.  in general that small savings is insignificant in larger code and also it means little unless you trying to cram it into a small eprom.

Allison

**Comment by hsoft** [comment]
> unless you trying to cram it into a small eprom.

it is **exactly** the kind of scenarios that Collapse OS aim for.

**Comment by Allisontheolder** [comment]
You missed that refactoring the code could save a lot more...

I'm old enough to have remember that 2858 (1k) or 2716(2K) was big and 
I still have a stock of them in metal tubes.  Generally tricks are good but 
usual code can be more compact.

**Comment by hsoft** [comment]
Of course, that goes without saying that there are many refactoring targets that are more significant than CALL -> RST. I'm not replying to that. I'm replying to "it means little unless"

**Comment by hsoft** [comment]
I tried to implement it and it took me a while to realize that there's a major roadblock here: During first pass, we compute labels offsets and we, of course, don't know yet the value of that label. This means that we cannot know during first pass if we are going to spit a CALL or a RST. But because we can't know, it means that we can't have the proper binary offset count either: If a particular call ends up being a RST and that my counter goes up 3 bytes instead of 1, then all my labels offsets are wrong.

Maybe that through deep refactoring, this would become possible, but it's definitely not worth it. Closing.
